{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/mulqueenr/DSWG_MethylationWorkshop/blob/main/DSWG_Methylation.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "u2Aw888UA5Tk"
      },
      "source": [
        "# Welcome to the DSWG Methylation workshop!\n",
        "\n",
        "I'll be running you through an vignette of amethyst processing, based heavily off the vignette for [PBMC processing available on the amethyst github repo.](https://htmlpreview.github.io/?https://github.com/lrylaarsdam/amethyst/blob/main/vignettes/pbmc_vignette/pbmc_vignette.html)\n",
        "\n",
        "We will be processing from breast cancer cell line cells: [MDA-MB-231](https://en.wikipedia.org/wiki/MDA-MB231), [MCF7](https://en.wikipedia.org/wiki/MCF-7), and [MCF10A](https://www.cellosaurus.org/CVCL_0598)\n",
        "\n",
        "Our goals in this workshop are to:\n",
        "\n",
        "1. Get introduced to data structures of amethyst.\n",
        "2. Get familiar with quality metrics of DNA methylation.\n",
        "3. Cluster single cell methylomes to separate out the cell lines.\n",
        "4. (optional) Perform copy number calling to confirm our clusters."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5rkLePAHBUL4"
      },
      "source": [
        "## Library Installation\n",
        "\n",
        "I prebuilt some R libraries that we can use to skip the install process. If you want to run the install yourself, the code is commented out below. However, you can fully ignore the commented out code below. All it is showing is that we need to install quite a few dependencies to get to the installation of two packages that will be the focus on this workshop:\n",
        "\n",
        "- **amethyst**: single cell methylation processing\n",
        "- **copykit**: single cell whole genome copy number calling"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "d8jGVU8PBf2S"
      },
      "outputs": [],
      "source": [
        "#first lets install packages for the tutorial\n",
        "# this is prerun and stored on google drive to save time\n",
        "\n",
        "#install.packages(\"BiocManager\")\n",
        "\n",
        "#library(\"BiocManager\")\n",
        "#BiocManager::install(c(\"caret\", \"devtools\", \"data.table\", \"dplyr\", \"furrr\", \"future\", \"future.apply\",\n",
        "#  \"ggplot2\", \"grDevices\", \"gridExtra\", \"igraph\", \"irlba\", \"janitor\", \"Matrix\", \"methods\", \"pheatmap\",\n",
        "#  \"plotly\", \"plyr\", \"purrr\", \"randomForest\", \"rhdf5\", \"rtracklayer\", \"Rtsne\", \"scales\", \"stats\", \"stringr\",\n",
        "#  \"tibble\", \"tidyr\", \"umap\", \"utils\"))\n",
        "\n",
        "#library(\"devtools\")\n",
        "\n",
        "#devtools::install_github(\"JinmiaoChenLab/Rphenograph\")\n",
        "#devtools::install_github(\"KrishnaswamyLab/MAGIC/Rmagic\")\n",
        "#devtools::install_github(\"TomKellyGenetics/leiden\")\n",
        "#installing older version of amethyst for compatibility with cell line data files.\n",
        "#system('wget https://github.com/lrylaarsdam/amethyst/releases/download/v0.0.0.9000/amethyst_0.0.0.9000.tar.gz')\n",
        "#system('tar -xf amethyst_0.0.0.9000.tar.gz')\n",
        "#devtools::install(\"/content/amethyst\")\n",
        "#devtools::install_github(\"navinlabcode/copykit\")\n",
        "#system('cd /usr/local/lib/R/site-library/; tar -czf /content/R_library.tar.gz *; cd /content')\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zCim33NdCT6w"
      },
      "source": [
        "To save ~30min of installation time, I prebuild the library and uploaded to a google drive folder. The following code downloads the libraries, and uncompresses them, so we can use them in our R session.\n",
        "\n",
        "\n",
        "> *system()* takes a string and calls it in the terminal above. Likewise, we can run these string commands using the terminal button on the bottom left of colab.\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "-QpeUSLQ_od-"
      },
      "outputs": [],
      "source": [
        "#download prebuilt R libraries\n",
        "system(\"gdown '1r2P-GZa2QukEl-i619yK7dqeoEhGjBNN'\")\n",
        "system(\"mkdir R_library\")\n",
        "system(\"tar -xvf R_library.tar.gz -C R_library\") #unpack the library"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "We'll also download the cell line methylation data while here."
      ],
      "metadata": {
        "id": "cww28JQ7OuRM"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ni42_iGg_EG6"
      },
      "outputs": [],
      "source": [
        "#download data\n",
        "system(\"gdown '195TI6-r2ruEGaKK4rIfWa24aT2xyEjL5'\")\n",
        "system(\"tar -xf workshop_data.tar.gz\") #unpack the library"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "M_BmbTmuHuL9"
      },
      "source": [
        "If you check your files for this colab session, you should see them populate with two folders:\n",
        "\n",
        "- **R_library.tar.gz** : this is our downloaded compressed libraries\n",
        "- **/R_library/** : this is our decompressed folder\n",
        "- **h5_files** : this contains our methylation data in compressed h5 format\n",
        "- **metadata.tsv** : this is a tab-separated format of our metadata\n",
        "- **workshop.copykit.obj.rds** : this is a bonus copykit object that we will use for cnv calling\n",
        "\n",
        "Now all we have to do is tell R where to look for these libraries and we should be good to go!\n",
        "\n",
        "To do this, we are going to append the folder location to an R environment variable called \".libPaths\".\n",
        "\n",
        "Following this we will load our packages.\n",
        "\n",
        "\n",
        "\n",
        "> *library()* loads packages into the R runtime.\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "VzVqptnMAEY8"
      },
      "outputs": [],
      "source": [
        ".libPaths(c(\"/content/R_library\", .libPaths()))\n",
        "library(amethyst) #for methylation analysis\n",
        "library(copykit) #for cnv calling\n",
        "library(ggplot2) #for plotting\n",
        "library(cowplot) #for combining plots\n",
        ".libPaths()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "d8rGjJoWL6ZM"
      },
      "source": [
        "# Let's Assemble our Amethyst Object\n",
        "\n",
        "![amethyst_layout.png](https://github.com/lrylaarsdam/amethyst/blob/main/images/objectstructure.png?raw=true)"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Read in our data\n"
      ],
      "metadata": {
        "id": "3ifHYacxP74l"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "> *read.csv* reads in a file, by default a comma separated value (csv) file. We are going to specify that the separators are tabs (a tsv) file.\n",
        "\n"
      ],
      "metadata": {
        "id": "zRHGl09HQa90"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "2yFbgcEvAnTB"
      },
      "outputs": [],
      "source": [
        "meta<-read.csv(file=\"/content/metadata.tsv\",header=T,sep=\"\\t\")\n",
        "head(meta)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "rUKreygCMkub"
      },
      "outputs": [],
      "source": [
        "#read in our h5 filesobj <- addCellInfo(obj, file = \"metadata.tsv\")\n",
        "h5_paths<-read.csv(file=\"/content/h5_files/h5_paths.tsv\",header=T,sep=\"\\t\")\n",
        "h5_paths$barcodes<-row.names(h5_paths)\n",
        "h5_paths$paths<-h5_paths$path\n",
        "h5_paths<-h5_paths[,c(\"barcodes\",\"paths\")]\n",
        "head(h5_paths)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sEzIsz_EZdi1"
      },
      "source": [
        "## Create an empty Amethyst Object\n",
        "\n",
        "You'll notice for both the h5 paths and the metadata, the row names of the files are the same. So through this we will like our cell info to the methylation profiles! Now let's make a new amethyst object.\n",
        "\n",
        "\n",
        "\n",
        "> *str()* calls the \"structure\" of an object. This can be a simple type (i.e. character or numeric) or can show the hierarchical layout of a fancier structure. Its great practice to know how and where your data is stored in an object!\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ro7fnFXfZmaB"
      },
      "outputs": [],
      "source": [
        "obj <- createObject()\n",
        "str(obj)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BIyA-t2ZZufn"
      },
      "source": [
        "The \"@\" symbol shows a slot in the object. These are predefined in R, so they only accept certain data types and formats."
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Add our data to the object\n",
        "\n",
        "We will start by adding our metadata into the @metadata slot and h5 paths into the @h5paths slot. Then we will add reference data using a handy wrapper function for amethyst. This will give us gene locations along the genome, which will be useful for plotting later.\n",
        "\n",
        "We'll then just do another quick check to make sure it got added correctly."
      ],
      "metadata": {
        "id": "b2uhpN66RKCD"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "gaA90NcOZ8ud"
      },
      "outputs": [],
      "source": [
        "obj@metadata<-meta\n",
        "obj@h5paths<-h5_paths[row.names(meta),]"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PMe_rFIjaScL"
      },
      "source": [
        "R structured objects are hierarchical. So we can call a slot, and then call further objected within that slot.\n",
        "\n",
        "Let's make a quick counts table of our \"sample\" metadata column."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "eQdi4F_7aSzP"
      },
      "outputs": [],
      "source": [
        "table(obj@metadata$sample)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hGi1iK0Pa4IM"
      },
      "source": [
        "### QC Plot: All Cytosines covered\n",
        "\n",
        "Looks like we're working with about 50 cells a-piece for the 3 different cell lines. Let's take a peek at the cytosine coverage per cell as a histogram."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "2tsswidpavli"
      },
      "outputs": [],
      "source": [
        "ggplot(obj@metadata, aes(x = cov)) +\n",
        "  geom_histogram(bins = 100) +\n",
        "  theme_classic()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "YXz-Vc6fbUTL"
      },
      "source": [
        "### QC Plot: CGs covered\n",
        "\n",
        "And now just the CG dinucleotide coverage. By Calling a similar function, but just changing the metadata column."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lIo9HltXbc1z"
      },
      "outputs": [],
      "source": [
        "ggplot(obj@metadata, aes(x = cg_cov)) +\n",
        "  geom_histogram(bins = 100) +\n",
        "  theme_classic()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "fao3bwdrcY50"
      },
      "source": [
        "### Filtering on CGs covered\n",
        "Lets just filter out data a bit to remove those high and low coverage cells. This will help reduce technical bias in our analysis. Let's define a threshold and plot it over our histogram, just to visually check. And then lets apply these thresholds as a filter for our cells."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "RZ-RgUkfcbsb"
      },
      "outputs": [],
      "source": [
        "ggplot(obj@metadata, aes(x = cg_cov)) +\n",
        "  geom_histogram(bins = 100) +\n",
        "  theme_classic() +\n",
        "  geom_vline(xintercept=c(100000,500000),color=\"red\")"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Here we are combining a couple elements of R programming.\n",
        "\n",
        "First we will use the data structure hierarchy to call the metadata column we want:\n",
        "\n",
        "\n",
        "```\n",
        "obj@metadata$cg_cov\n",
        "```\n",
        "\n",
        "Then we will check if that value, per cell is greater than 100,000:\n",
        "\n",
        "\n",
        "\n",
        "```\n",
        "obj@metadata$cg_cov>100000\n",
        "```\n",
        "\n",
        "This gives us a list, the same length as obj@metadata$cg_cov of just TRUE or FALSE. For if that cell does pass that filter. We can apply the second threshold to make two lists of TRUE or FALSE and add them together, so only cells that are TRUE in both lists are maintained as TRUE.\n",
        "\n",
        "\n",
        "\n",
        "```\n",
        "obj@metadata$cg_cov>100000 & obj@metadata$cg_cov<500000\n",
        "```\n",
        "\n",
        "Finally, we use the function\n",
        "\n",
        "\n",
        "> *which()* This simply takes all the TRUE values in the list, and gives us their index, which is equivalent to their row number.\n",
        "\n",
        "So from this, we can subset the obj@metadata data frame and get all the passing cell IDs in a single line of strung together code.\n",
        "\n",
        "Finally, we will use a amethyst function:\n",
        "\n",
        "\n",
        "\n",
        "> *subsetObject()* takes a list of cells and removes all others.\n",
        "\n",
        "This is convenient for us, because it will look through all the slots in one call.\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "W40pR4m0Rr22"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7hGVfMZgcuIu"
      },
      "outputs": [],
      "source": [
        "#first we make a list of cells that are within the thresholds\n",
        "cells_pass_qc<-row.names(obj@metadata[which(obj@metadata$cg_cov>100000 & obj@metadata$cg_cov<500000),])\n",
        "\n",
        "#then we use the subset Object function to keep only passing cells\n",
        "obj<-subsetObject(obj,cells=cells_pass_qc)\n",
        "\n",
        "#a final check of cell counts per sample\n",
        "table(obj@metadata$sample)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GzlU5S6Pbr3N"
      },
      "source": [
        "### QC Plot: Percent mCG\n",
        "\n",
        "And lets check the percent methylation. For most cells this ranging from about 60% to 80% depending on cell type. Cancer and cell lines tend to be on the lower end generally."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "6prYS9txaRts"
      },
      "outputs": [],
      "source": [
        "ggplot(obj@metadata, aes(x = sample,y=mcg_pct)) +\n",
        "  geom_boxplot() +\n",
        "  theme_classic() +\n",
        "  ylim(c(0,100))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xgu30UJFbEVD"
      },
      "source": [
        "# Indexing and Methylation Windows\n",
        "\n",
        "### Indexing H5\n",
        "\n",
        "First we will index our CG sites in our h5 files. This will allow us to access them for summarizing over windows with much improved memory efficiency.\n",
        "\n",
        "\n",
        "\n",
        "> *indexChr()* generates indexes for h5 files, so the whole file does not need to be stored in memory.\n",
        "\n",
        "This takes a few min to run...\n",
        "\n",
        "\n",
        "We can now take a peek at some chr1 data."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "B2tuyTDs8tDc"
      },
      "outputs": [],
      "source": [
        "obj@index[[\"chr_cg\"]] <- indexChr(obj, type = \"CG\", threads = 3)\n",
        "head(obj@index[[\"chr_cg\"]][[\"chr1\"]])"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Wq3d5JrDU58B"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GMVxxt_1diqC"
      },
      "source": [
        "### Filtering out Scaffolds\n",
        "Next let's filter down to just chromosomes. Reference genomes tend to have excess contigs/scaffolds that are gene-poor and lower quality mapping regions.\n",
        "\n",
        "Similar to how we filtered cells passing a CG count threshold, we can filter chromosomes just based on name.\n",
        "\n",
        "To generate our TRUE/FALSE list I am using the function:\n",
        "\n",
        "\n",
        "> *startsWith()* This function takes a string (our contig) and checks if it starts with a given prefix, in this case (\"chr\")\n",
        "\n",
        "We **negate** this result with a **!** meaning, contigs that **DON'T** start with \"chr\" are passed on to our if statement.\n",
        "\n",
        "We then take those non-chromosome contigs and remove them by resetting them as \"NULL\" in our object.\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "cBpZBAG-QSpo",
        "collapsed": true
      },
      "outputs": [],
      "source": [
        "#filter to just autosomes\n",
        "chr_names<-names(obj@index[[\"chr_cg\"]])\n",
        "print(\"All contigs:\")\n",
        "print(chr_names)\n",
        "\n",
        "for(contig in chr_names){\n",
        "  if(!startsWith(x=contig,prefix=\"chr\")){\n",
        "    obj@index[[\"chr_cg\"]][[contig]]<-NULL\n",
        "  }\n",
        "}\n",
        "\n",
        "chr_names<-names(obj@index[[\"chr_cg\"]])\n",
        "print(\"Chromosomes passing filter:\")\n",
        "print(chr_names)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "NFj1RCSjd-li"
      },
      "source": [
        "### Summarizing Methylation Signal over Genomic Windows\n",
        "\n",
        "Single cell methylation data is **sparse**. Meaning the likelihood that we capture the same CG site across many cells is low. Our data has roughly 200k CGs covered per cell, but thats just a fraction of the 28M in the human genome. To get around this sparsity, we use larger windows that are biologically co-regulated, these are generally 500bp-5kbp in size.\n",
        "\n",
        "For our workshop, we are going to use 5Mbp windows, which are huge! But these cell lines should be different enough to still cluster. For real analysis you will need a high powered computing cluster to calculate smaller windows for more cells.\n",
        "\n",
        "Also note, we are using \"**score**\" as a metric rather than \"**percent**\".\n",
        "\n",
        "This is because we are correcting for differences in coverage per cell across windows. Basically the more reads we have in a window, the more confident we are in the methylation signature there.\n",
        "\n",
        "This takes about 5 to 6min..."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "JB7G0mJk_Q7a"
      },
      "outputs": [],
      "source": [
        "obj@genomeMatrices[[\"cg_5mb_score\"]] <- makeWindows(obj,\n",
        "                                                    stepsize = 5000000,\n",
        "                                                    type = \"CG\",\n",
        "                                                    metric = \"score\",\n",
        "                                                    threads = 4,\n",
        "                                                    index = \"chr_cg\",\n",
        "                                                    nmin = 2)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "25QfITtdfaTz"
      },
      "source": [
        "### Filtering Windows\n",
        "\n",
        "We can also filter windows that don't have good coverage. Methylation profiling methods rely on inducing point mutations and creates sites that are difficult to map. We can filter those sites out."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "xX09LgxuTzT-"
      },
      "outputs": [],
      "source": [
        "#filter to windows with 90% of cells have data\n",
        "obj@genomeMatrices[[\"cg_5mb_score\"]] <- obj@genomeMatrices[[\"cg_5mb_score\"]][rowSums(!is.na(obj@genomeMatrices[[\"cg_5mb_score\"]])) >= nrow(obj@metadata)*.9, ]\n",
        "\n",
        "win_count<-nrow(obj@genomeMatrices[[\"cg_5mb_score\"]])\n",
        "print(paste(win_count,\"windows passed filter.\"))\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mYoVUAbkfuZd"
      },
      "source": [
        "# Clustering\n",
        "\n",
        "Now that we have our summarized windows, we can perform clustering.\n",
        "\n",
        "Methylation, like RNA or ATAC data is clustered by first using a dimensionality reduction (PCA, IRLBA, LSI) to reduce features while accounting for feature-feature correlations.\n",
        "\n",
        "We will use two functions in this block:\n",
        "\n",
        "\n",
        "\n",
        "> *dimEstimate()* takes in the matrix and estimates the top variable components to a certain threshold.\n",
        "\n",
        "> *runIrlba()* runs the dimensionality reduction, using a PCA that can handle NA values.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "36xu7S5uUFuY"
      },
      "outputs": [],
      "source": [
        "#can run a dimensionality estimate to see how many principal components are informative\n",
        "dimEstimate(obj, genomeMatrices = c(\"cg_5mb_score\"), dims = c(10), threshold = 0.95)\n",
        "\n",
        "set.seed(111)\n",
        "\n",
        "# name the result whatever you want. I like descriptive names, at the cost of length.\n",
        "# run a modified PCA (IRLBA) which accounts for NA values\n",
        "obj@reductions[[\"irlba_cg_5mb_score\"]] <- runIrlba(obj, genomeMatrices = c(\"cg_5mb_score\"), dims = c(5), replaceNA = c(0))\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Because our methylation values per window is weighed by confidence, we have some principal components which may have a strong coverage bias. I.e. they are capturing technical artifacts rather than real biology. To fix this, we can regress out the coverage per each PC.\n",
        "\n",
        " We will use\n",
        "\n",
        "\n",
        "> *regressCovBias()* to do this. It will look for our metadata column \"*cov*\" for regression.\n",
        "\n"
      ],
      "metadata": {
        "id": "QaMpXTFVhbVV"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "PUqBn9-vXJPa"
      },
      "outputs": [],
      "source": [
        "#regress out coverage across the PCs to mitigate sequencing and QC effects on biology\n",
        "obj@reductions[[\"irlba_cg_5mb_score_regressed\"]] <- regressCovBias(obj, reduction = \"irlba_cg_5mb_score\")"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "With our regressed PCs we can now run our clustering to group cells in similar space together.\n",
        "\n",
        "\n",
        "\n",
        "> *runCluster()* this will wrap Rphenograph, a function to group cells by louvain clustering.\n",
        "\n"
      ],
      "metadata": {
        "id": "b23ghHcQh95o"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "X4FzUL-mXL8e"
      },
      "outputs": [],
      "source": [
        "#cluster cells in shared space\n",
        "obj <- runCluster(obj, k=50, reduction = \"irlba_cg_5mb_score_regressed\")\n",
        "table(obj@metadata$cluster_id) #get a count of cells per cluster"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now that we have our clusters together, lets cram them into 2D space!\n",
        "\n",
        "\n",
        "\n",
        "> *amethyst::runUmap()* UMAP wrapper for amethyst, given our coverage regressed PCs, using euclidean distance.\n",
        "\n",
        "Notably, in this function call, we first define the package we want the function from \"amethyst::\" this is because copykit has a function called runUmap too! It is good form to identify the package for functions (but in most cases you don't have mutiple functions with the same name, so its okay).\n"
      ],
      "metadata": {
        "id": "bEmIWNuViVAd"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "NHX2venqXNNY"
      },
      "outputs": [],
      "source": [
        "#run umap to visualize in 2D\n",
        "obj <- amethyst::runUmap(obj, method = \"euclidean\", reduction = \"irlba_cg_5mb_score_regressed\")"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now let's plot! We will use the amethyst function:\n",
        "\n",
        ">*dimFeature()* to plot our umap reduction by various metadata columns.\n",
        "\n",
        "We can then plot all our plots together (p1,p2,p3,p4) for ease of visualization using the library cow_plot's *plot_grid()* function.\n",
        "\n"
      ],
      "metadata": {
        "id": "e38HeZGljnXy"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lgxwwZ6PXOv1"
      },
      "outputs": [],
      "source": [
        "#plot by multiple metadata columns for visualization\n",
        "print(\"Plotting...\")\n",
        "p1 <- dimFeature(obj, colorBy = cluster_id, reduction = \"umap\") + ggtitle(\"Clusters\")\n",
        "p2 <- dimFeature(obj, colorBy = sample, reduction = \"umap\") + ggtitle(\"Samples\")\n",
        "p3 <- dimFeature(obj, colorBy = log10(cov), pointSize = 1) + scale_color_gradientn(colors = c(\"black\", \"turquoise\", \"gold\", \"red\"),guide=\"colourbar\") + ggtitle(\"Coverage distribution\")\n",
        "p4 <- dimFeature(obj, colorBy = mcg_pct, pointSize = 1) + scale_color_gradientn(colors = c(\"black\", \"turquoise\", \"gold\", \"red\")) + ggtitle(\"Global %mCG distribution\")\n",
        "plt<-plot_grid(p1, p2,p3, p4,ncol=2)\n",
        "plt"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's compare our cell line metadata in \"sample\" to our assigned \"cluster_id\" from the *runCluster* function above."
      ],
      "metadata": {
        "id": "MCXyf3mfkC2E"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "PjLU1Kc5XhLm"
      },
      "outputs": [],
      "source": [
        "sampleComp(obj, groupBy = \"sample\", colorBy = \"cluster_id\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-de-Kg3Qh1tt"
      },
      "source": [
        "Remember! This is all being done on 5mb windows, which are about 1000x larger than standard methylome processing!\n",
        "\n",
        "We can further define windows to just those that show high variance, those significantly differentially methylated, or promoters of known RNA marker genes.\n",
        "\n",
        "I pre-baked the methylation summarized over much smaller windows, which we can use to visualize genomic regions in more detail. Let's download it and load it into our amethyst object."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "-f4_lH7OhhD-"
      },
      "outputs": [],
      "source": [
        "system(\"gdown '1f6OdHmrR24K337_RX1x1eSgC_QrcLWWd'\") #download it\n",
        "tracks<-readRDS(\"celltype_500bpwindow_tracks.rds\") #load it into R\n",
        "obj@genomeMatrices[[\"cg_cluster_tracks\"]] <- tracks[[\"pct_matrix\"]] #load it into amethyst object\n",
        "obj@ref <- makeRef(\"hg38\")"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's use two other amethyst plotting functions to look at methylation over some genes of interest!\n",
        "\n",
        "Let's compare and contrast using *heatMap()* and *histograM()* for visualization. Which do you think is more clear?\n",
        "\n"
      ],
      "metadata": {
        "id": "yfIlxH5Qkcz2"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Ao-fdKQMiDAq"
      },
      "outputs": [],
      "source": [
        "amethyst::heatMap(obj,\n",
        "        genes = c(\"ERBB2\"),\n",
        "        matrix = \"cg_cluster_tracks\",\n",
        "        legend = T)\n",
        "\n",
        "amethyst::heatMap(obj,\n",
        "        genes = c(\"PGR\"),\n",
        "        matrix = \"cg_cluster_tracks\",\n",
        "        legend = T)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "GnDWviOd1Rsh"
      },
      "outputs": [],
      "source": [
        "histograM(obj,\n",
        "          baseline = \"mean\",\n",
        "          genes = \"PGR\",\n",
        "          matrix = \"cg_cluster_tracks\",\n",
        "          legend = F)"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Conclusion\n",
        "\n",
        "So we went from compressed files of chromosome, location, methylation for about 150 cells all the way to UMAP clustering, and genomic window visualization! Not bad for an hour!\n",
        "\n",
        "Of course this is plenty more you can do with methylation data, and this just serves as an intro!\n",
        "\n",
        "Here's some other topics with existing packages to explore:\n",
        "\n",
        "\n",
        "*   Identification of differentially methylated regions (DMRs)\n",
        "*   Methylation based lineage analysis via [MethylTree](https://github.com/ShouWenWang-Lab/MethylTree/tree/master)\n",
        "* Copy number calling by alignment rates across the genome via [copykit](https://github.com/navinlabcode/copykit)\n",
        "* Methylation clock analysis for relative sample and cell type aging via [scAGE](https://github.com/alex-trapp/scAge)\n",
        "\n",
        "Amethyst has plenty more resources and functions we didn't touch on! Please feel free to peruse their [vignettes](https://github.com/lrylaarsdam/amethyst) and example data for more!"
      ],
      "metadata": {
        "id": "6C5oiozbkt97"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Bonus Sections\n",
        "\n",
        "I included some bonus sections, time allowing, that we can delve into a bit deeper."
      ],
      "metadata": {
        "id": "2aVE998BlmZV"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Bonus: Calculating DMRs\n",
        "\n",
        "Like we limited our clustering windows to just chromosomes. We can also limit our DMR analysis to just chromosome 1. This is just to speed things up for the workshop."
      ],
      "metadata": {
        "id": "28Zndq1GOZN2"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#lets subset our 5kb matrix to just chr1\n",
        "chr_names<-names(tracks[[\"sum_matrix\"]])\n",
        "for(contig in chr_names){\n",
        "  if(!(contig==\"chr1\")){\n",
        "    obj@index[[\"chr_cg\"]][[contig]]<-NULL\n",
        "  }\n",
        "}"
      ],
      "metadata": {
        "id": "5EMUUfZHUP3J"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "With that filtered, we can now run amethyst's *testDMR()* function. This takes in our 5kb matrix (which now just has chr1) and does a one by rest comparison for each cell line."
      ],
      "metadata": {
        "id": "tURjOqdWUSQz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#takes about 10min\n",
        "dmrs <- testDMR(tracks[[\"sum_matrix\"]], # Sum of c and t observations in each genomic window per group\n",
        "                eachVsAll = TRUE, # If TRUE, each group found in the sumMatrix will be tested against all others\n",
        "                nminTotal = 5, # Min number observations across all groups to include the region in calculations\n",
        "                nminGroup = 5) # Min number observations across either members or nonmembers to include the region in calculations"
      ],
      "metadata": {
        "id": "Deet_OJkUiDh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "This reports all values for all windows in all comparisons. So let's filter to those with a Bonferroni adjusted p value of <0.01 and a log threshold of 1.5x increase."
      ],
      "metadata": {
        "id": "Wk8QnfmiUkHU"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dmrs <- filterDMR(dmrs,\n",
        "                method = \"bonferroni\", # c(\"holm\", \"hochberg\", \"hommel\", \"bonferroni\", \"BH\", \"BY\", \"fdr\")\n",
        "                filter = TRUE, # If TRUE, removes insignificant results\n",
        "                pThreshold = 0.01, # Maxmimum adjusted p value to allow if filter = TRUE\n",
        "                logThreshold = 1.5) # Minimum absolute value of the log2FC to allow if filter = TRUE\n",
        "head(dmrs)"
      ],
      "metadata": {
        "id": "IiS-KNg0VNjn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "You may notice that \"test\" column has values 1,2 or 3 instead of our cell line groupings. We can rename this for plotting."
      ],
      "metadata": {
        "id": "BQMVpBNtVOLn"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dmrs$type <- dplyr::recode(dmrs$test,\n",
        "                                    \"1\" = \"MCF10A\",\n",
        "                                    \"2\" = \"MCF7\",\n",
        "                                    \"3\" = \"MDA-MB-231\")\n",
        "\n",
        "head(dmrs)"
      ],
      "metadata": {
        "id": "xctygJHGVXpi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Finally lets get a sense of how many DMRs we see on chr1 per cell line. Let's summarize our DMRs into counts, per group (type) and per direction (hypo or hypermethylated).\n",
        "\n",
        "Then lets plot these as bars."
      ],
      "metadata": {
        "id": "AxlYxZVVVaJM"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "iyVFXvVB3Q-U"
      },
      "outputs": [],
      "source": [
        "summarize_dmrs<-dmrs |> dplyr::group_by(type, direction) |> dplyr::summarise(n = dplyr::n())\n",
        "ggplot(summarize_dmrs,aes(y = type, x = n, fill = type)) +\n",
        "      geom_col() +\n",
        "      facet_grid(vars(direction), scales = \"free_y\") +\n",
        "      theme_classic()"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "You may notice that some of these DMRs are right next to eachother on the genome, and are thus likely co-regulated. We can collapse these together by distance. But this exercise is left to the reader."
      ],
      "metadata": {
        "id": "GVo6r03eVuR5"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "s_vx2JRSogFL"
      },
      "source": [
        "## Bonus: Copy number calling\n",
        "\n",
        "Because we have whole genome data, we can perform copy number calling just as we would with scWGS data.\n",
        "\n",
        "Please refer to the [copykit user guide](https://navinlabcode.github.io/CopyKit-UserGuide/index.html) for a lot more information!\n",
        "\n",
        "\n",
        "This bonus section will just be a brief coding section! But function descriptions and theory are available in the user guide."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "LXMMIwUrptYL"
      },
      "outputs": [],
      "source": [
        "#install one more package I forgot\n",
        "BiocManager::install(\"GenomeInfoDb\")\n",
        "library(BiocParallel)\n",
        "register(MulticoreParam(progressbar = T, workers = 2), default = T)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "kWqOQeCDoeLx"
      },
      "outputs": [],
      "source": [
        "cnv<-readRDS(\"workshop.copykit.obj.rds\")\n",
        "cnv"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "I97x8Yw6ospa"
      },
      "source": [
        "If we call the structure, you can see the hierarchical layout of a copykit object."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "p-0k_obYoxQn"
      },
      "outputs": [],
      "source": [
        "str(cnv)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#add in out cell line and methylation clustering data\n",
        "colData(cnv)$cellline<-NA\n",
        "colData(cnv)[row.names(obj@metadata),]$cellline<-obj@metadata$sample\n",
        "colData(cnv)$met_cluster<-NA\n",
        "colData(cnv)[row.names(obj@metadata),]$met_cluster<-obj@metadata$cluster_id\n",
        "\n",
        "table(colData(cnv)$cellline)\n",
        "table(obj@metadata$sample)"
      ],
      "metadata": {
        "id": "RMthv26Q8P4q"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's normalize the windows with a variance stabilized transformation."
      ],
      "metadata": {
        "id": "VMdM2vWIerYk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cnv<-runVst(cnv)"
      ],
      "metadata": {
        "id": "xHrV7vbStqO7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#takes about 10min currently\n",
        "cnv<-suppressWarnings(runSegmentation(cnv))"
      ],
      "metadata": {
        "id": "ZitAB1SYttTV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "cnv<-logNorm(cnv)"
      ],
      "metadata": {
        "id": "185wA5n_xHvI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#takes ~25min\n",
        "#cnv <- suppressWarnings(knnSmooth(cnv))"
      ],
      "metadata": {
        "id": "Wz3f-PI6xJaN"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "cnv <- copykit::runUmap(cnv)"
      ],
      "metadata": {
        "id": "BjimW1_i2LT_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "cnv <- findSuggestedK(cnv)\n",
        "cnv <- findClusters(cnv)"
      ],
      "metadata": {
        "id": "MwDzDTBr2Muw"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "p1<-plotUmap(cnv, label = 'cellline')\n",
        "p2<-plotUmap(cnv, label = 'subclones')\n",
        "p3<-plotUmap(cnv, label = 'met_cluster')\n",
        "plt<-plot_grid(p1, p2, p3)\n",
        "print(plt)"
      ],
      "metadata": {
        "id": "2FRKQo172N3X"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "ZIpNgYN6sDaQ"
      },
      "outputs": [],
      "source": [
        "\n",
        "plotHeatmap(\n",
        "  cnv,\n",
        "  label = c('cellline','met_cluster','subclones'),\n",
        "  genes = c(\"TP53\", \"BRAF\", \"MYC\"))\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyNal+O8dqpJWPiTpQPdngCS",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "R",
      "name": "ir"
    },
    "language_info": {
      "name": "R"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}