{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyNHbHaFuN4ocvCO1U6IlNGs",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/mulqueenr/DSWG_MethylationWorkshop/blob/main/DSWG_Methylation.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Welcome to the DSWG Methylation workshop!\n",
        "\n",
        "I'll be running you through an vignette of amethyst processing, based heavily off the vignette for PBMC processing available on the amethyst github repo.\n",
        "\n",
        "We will be processing from breast cancer cell line cells.\n",
        "\n",
        "Our goals in this workshop are to:\n",
        "\n",
        "- Get familiar with quality metrics of DNA methylation\n",
        "- Cluster single cell methylomes to separate out the cell lines.\n",
        "- Perform copy number calling to confirm our clusters."
      ],
      "metadata": {
        "id": "u2Aw888UA5Tk"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Library Installation\n",
        "\n",
        "I prebuilt some R libraries that we can use to skip the install process. If you want to run the install yourself, the code is commented out below. You can fully ignore the commented out code below. All it is showing is that we need to install quite a few dependencies to get to the installation of two packages that will be the focus on this workshop:\n",
        "\n",
        "- **amethyst**: single cell methylation processing\n",
        "- **copykit**: single cell whole genome copy number calling"
      ],
      "metadata": {
        "id": "5rkLePAHBUL4"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#first lets install packages for the tutorial\n",
        "#if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
        "#install.packages(\"BiocManager\")\n",
        "\n",
        "#library(\"BiocManager\")\n",
        "#BiocManager::install(c(\"caret\", \"devtools\", \"data.table\", \"dplyr\", \"furrr\", \"future\", \"future.apply\",\n",
        "#  \"ggplot2\", \"grDevices\", \"gridExtra\", \"igraph\", \"irlba\", \"janitor\", \"Matrix\", \"methods\", \"pheatmap\",\n",
        "#  \"plotly\", \"plyr\", \"purrr\", \"randomForest\", \"rhdf5\", \"rtracklayer\", \"Rtsne\", \"scales\", \"stats\", \"stringr\",\n",
        "#  \"tibble\", \"tidyr\", \"umap\", \"utils\"))\n",
        "\n",
        "#library(\"devtools\")\n",
        "\n",
        "#devtools::install_github(\"JinmiaoChenLab/Rphenograph\")\n",
        "#devtools::install_github(\"KrishnaswamyLab/MAGIC/Rmagic\")\n",
        "#devtools::install_github(\"TomKellyGenetics/leiden\")\n",
        "#devtools::install_github(\"lrylaarsdam/amethyst\")\n",
        "#devtools::install_github(\"navinlabcode/copykit\")"
      ],
      "metadata": {
        "id": "d8jGVU8PBf2S"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "To save ~30min of installation time, I prebuild the library and uploaded to a google drive folder. The following code downloads the libraries, and uncompresses them, so we can use them in our R session."
      ],
      "metadata": {
        "id": "zCim33NdCT6w"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-QpeUSLQ_od-",
        "outputId": "c69e3d3e-c8d4-4012-c6e9-0a713b210688"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Downloading...\n",
            "From (original): https://drive.google.com/uc?id=1r2P-GZa2QukEl-i619yK7dqeoEhGjBNN\n",
            "From (redirected): https://drive.google.com/uc?id=1r2P-GZa2QukEl-i619yK7dqeoEhGjBNN&confirm=t&uuid=3f591552-096d-4863-a51e-9f99707b1e8e\n",
            "To: /content/R_library.tar.gz\n",
            "100% 600M/600M [00:04<00:00, 150MB/s]\n"
          ]
        }
      ],
      "source": [
        "#download prebuilt R libraries\n",
        "!gdown 1r2P-GZa2QukEl-i619yK7dqeoEhGjBNN\n",
        "!tar -xf R_library.tar.gz #unpack the library\n",
        "#download data\n",
        "#!gdown\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "If you check your files for this colab session, you should see them populate with two folders:\n",
        "\n",
        "- R_library.tar.gz : this is our downloaded compressed libraries\n",
        "- content : this is our decompressed folder\n",
        "\n",
        "Now all we have to do is tell R where to look for these libraries and we should be good to go!\n",
        "\n",
        "To do this, we are going to append the folder location to an R environment variable called \".libPaths\".\n",
        "\n",
        "Following this we will load the two packages."
      ],
      "metadata": {
        "id": "r3tJ-MB6DDOU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        ".libPaths(c(\"/content/content/library\", .libPaths()))\n",
        "library(amethyst)\n",
        "library(copykit)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "VzVqptnMAEY8",
        "outputId": "db302625-e7e9-429a-983c-23103a3a4b26"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "    WARNING: The R package \"reticulate\" only fixed recently\n",
            "    an issue that caused a segfault when used with rpy2:\n",
            "    https://github.com/rstudio/reticulate/pull/1188\n",
            "    Make sure that you use a version of that package that includes\n",
            "    the fix.\n",
            "    Welcome to Amethyst! If you use Amethyst in your work, please cite our associated manuscript:\n",
            "      Rylaarsdam, L.E. et al. Single-cell DNA methylation analysis tool Amethyst resolves distinct non-CG methylation patterns in human astrocytes and oligodendrocytes.\n",
            "      Commun. Biol. 8, 1468 (2025). https://www.nature.com/articles/s42003-025-08859-2 \n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Attaching amethyst version 1.0.4.\n",
              "Loading required package: SingleCellExperiment\n",
              "Loading required package: SummarizedExperiment\n",
              "Loading required package: MatrixGenerics\n",
              "Loading required package: matrixStats\n",
              "\n",
              "Attaching package: ‘MatrixGenerics’\n",
              "\n",
              "The following objects are masked from ‘package:matrixStats’:\n",
              "\n",
              "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
              "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
              "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
              "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
              "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
              "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
              "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
              "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
              "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
              "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
              "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
              "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
              "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
              "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
              "    rowWeightedSds, rowWeightedVars\n",
              "\n",
              "Loading required package: GenomicRanges\n",
              "Loading required package: stats4\n",
              "Loading required package: BiocGenerics\n",
              "Loading required package: generics\n",
              "\n",
              "Attaching package: ‘generics’\n",
              "\n",
              "The following objects are masked from ‘package:base’:\n",
              "\n",
              "    as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,\n",
              "    setequal, union\n",
              "\n",
              "\n",
              "Attaching package: ‘BiocGenerics’\n",
              "\n",
              "The following objects are masked from ‘package:stats’:\n",
              "\n",
              "    IQR, mad, sd, var, xtabs\n",
              "\n",
              "The following objects are masked from ‘package:base’:\n",
              "\n",
              "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
              "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
              "    get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget,\n",
              "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
              "    rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique,\n",
              "    unsplit, which.max, which.min\n",
              "\n",
              "Loading required package: S4Vectors\n",
              "\n",
              "Attaching package: ‘S4Vectors’\n",
              "\n",
              "The following object is masked from ‘package:utils’:\n",
              "\n",
              "    findMatches\n",
              "\n",
              "The following objects are masked from ‘package:base’:\n",
              "\n",
              "    expand.grid, I, unname\n",
              "\n",
              "Loading required package: IRanges\n",
              "Loading required package: Seqinfo\n",
              "Loading required package: Biobase\n",
              "Welcome to Bioconductor\n",
              "\n",
              "    Vignettes contain introductory material; view with\n",
              "    'browseVignettes()'. To cite Bioconductor, see\n",
              "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
              "\n",
              "\n",
              "Attaching package: ‘Biobase’\n",
              "\n",
              "The following object is masked from ‘package:MatrixGenerics’:\n",
              "\n",
              "    rowMedians\n",
              "\n",
              "The following objects are masked from ‘package:matrixStats’:\n",
              "\n",
              "    anyMissing, rowMedians\n",
              "\n",
              "Loading required package: Rsubread\n",
              "Loading required package: DNAcopy\n",
              "\n",
              "Attaching package: ‘copykit’\n",
              "\n",
              "The following object is masked from ‘package:amethyst’:\n",
              "\n",
              "    runUmap\n",
              "\n",
              "In addition: Warning message:\n",
              "replacing previous import ‘GenomicRanges::reduce’ by ‘purrr::reduce’ when loading ‘amethyst’ \n"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "?runCountReads()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2yFbgcEvAnTB",
        "outputId": "05f38983-4eb1-43a3-9a06-bd5a05822f2b"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "File: /tmp/RtmpIqFP05/Rtxt24d5c8e3559\n",
            "runCountReads             package:copykit              R Documentation\n",
            "\n",
            "\n",
            "\n",
            "_\bA_\bl_\bi_\bg_\bn_\bs _\bt_\bh_\be _\br_\be_\ba_\bd_\bs _\bf_\br_\bo_\bm _\bt_\bh_\be _\bB_\bA_\bM _\bf_\bi_\bl_\be _\bt_\bo _\bt_\bh_\be _\bv_\ba_\br_\bi_\ba_\bb_\bl_\be _\bb_\bi_\bn_\bn_\bi_\bn_\bg _\bp_\bi_\bp_\be_\bl_\bi_\bn_\be.\n",
            "\n",
            "\n",
            "\n",
            "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
            "\n",
            "\n",
            "\n",
            "     runCountReads performs the variable binning (VarBin) algorithm to\n",
            "\n",
            "     a series of BAM files resulting from short-read sequencing.\n",
            "\n",
            "\n",
            "\n",
            "_\bU_\bs_\ba_\bg_\be:\n",
            "\n",
            "\n",
            "\n",
            "     runCountReads(\n",
            "\n",
            "       dir,\n",
            "\n",
            "       genome = c(\"hg38\", \"hg19\"),\n",
            "\n",
            "       resolution = c(\"220kb\", \"55kb\", \"110kb\", \"195kb\", \"280kb\", \"500kb\", \"1Mb\", \"2.8Mb\"),\n",
            "\n",
            "       remove_Y = FALSE,\n",
            "\n",
            "       min_bincount = 10,\n",
            "\n",
            "       is_paired_end = FALSE,\n",
            "\n",
            "       BPPARAM = bpparam()\n",
            "\n",
            "     )\n",
            "\n",
            "     \n",
            "\n",
            "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
            "\n",
            "\n",
            "\n",
            "     dir: A path for the directory containing BAM files from short-read\n",
            "\n",
            "          sequencing.\n",
            "\n",
            "\n",
            "\n",
            "  genome: Name of the genome assembly. Default: 'hg38'.\n",
            "\n",
            "\n",
            "\n",
            "resolution: The resolution of the VarBin method. Default: '220kb'.\n",
            "\n",
            "\n",
            "\n",
            "remove_Y: (default == FALSE) If set to TRUE, removes information from\n",
            "\n",
            "          the chrY from the dataset.\n",
            "\n",
            "\n",
            "\n",
            "min_bincount: A numerical indicating the minimum mean bin counts a cell\n",
            "\n",
            "          should have to remain in the dataset.\n",
            "\n",
            "\n",
            "\n",
            "is_paired_end: A boolean indicating if bam files are from single-read\n",
            "\n",
            "          or pair end sequencing.\n",
            "\n",
            "\n",
            "\n",
            " BPPARAM: A BiocParallelParam specifying how the function should be\n",
            "\n",
            "          parallelized.\n",
            "\n",
            "\n",
            "\n",
            "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
            "\n",
            "\n",
            "\n",
            "     ‘runCountReads’ takes as input duplicate marked BAM files from\n",
            "\n",
            "     whole genome sequencing and runs the variable binning pipeline\n",
            "\n",
            "     algorithm. It is important that BAM files are duplicate marked.\n",
            "\n",
            "     Briefly, the genome is split into pre-determined bins. The bin\n",
            "\n",
            "     size is controlled by the argument ‘resolution’. By using VarBin,\n",
            "\n",
            "     for a diploid cell, each bin will receive equal amount of reads,\n",
            "\n",
            "     controlling for mappability. A lowess function is applied to\n",
            "\n",
            "     perform GC correction across the bins. The argument ‘genome’ can\n",
            "\n",
            "     be set to 'hg38' or 'hg19' to select the scaffolds genome\n",
            "\n",
            "     assembly. The scaffolds are GenomicRanges objects Information\n",
            "\n",
            "     regarding the alignment of the reads to the bins and from the bam\n",
            "\n",
            "     files are stored in the #' ‘colData’. ‘min_bincount’ Indicates the\n",
            "\n",
            "     minimum mean bincount a cell must present to be kept in the\n",
            "\n",
            "     dataset. Cells with low bincounts generally present bin dropouts\n",
            "\n",
            "     due to low read count that will be poorly segmented.\n",
            "\n",
            "\n",
            "\n",
            "_\bV_\ba_\bl_\bu_\be:\n",
            "\n",
            "\n",
            "\n",
            "     A matrix of bin counts within the scCNA object that can be\n",
            "\n",
            "     accessed with ‘bincounts’\n",
            "\n",
            "\n",
            "\n",
            "     #' @references Navin, N., Kendall, J., Troge, J. et al. Tumour\n",
            "\n",
            "     evolution inferred by single-cell sequencing. Nature 472, 90–94\n",
            "\n",
            "     (2011). https://doi.org/10.1038/nature09807\n",
            "\n",
            "\n",
            "\n",
            "     Baslan, T., Kendall, J., Ward, B., et al (2015). Optimizing sparse\n",
            "\n",
            "     sequencing of single cells for highly multiplex copy number\n",
            "\n",
            "     profiling. Genome research, 25(5), 714–724.\n",
            "\n",
            "     https://doi.org/10.1101/gr.188060.114\n",
            "\n",
            "\n",
            "\n",
            "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
            "\n",
            "\n",
            "\n",
            "     Darlan Conterno Minussi\n",
            "\n",
            "\n",
            "\n",
            "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
            "\n",
            "\n",
            "\n",
            "     ## Not run:\n",
            "\n",
            "     \n",
            "\n",
            "     copykit_obj <- runCountReads(\"/PATH/TO/BAM/FILES\")\n",
            "\n",
            "     ## End(Not run)\n",
            "\n",
            "     \n",
            "\n",
            "---\n"
          ]
        }
      ]
    }
  ]
}