{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyN6kd/D5t5jGpx0ZHdOpuLC",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/mulqueenr/DSWG_MethylationWorkshop/blob/main/DSWG_Methylation.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Welcome to the DSWG Methylation workshop!\n",
        "\n",
        "I'll be running you through an vignette of amethyst processing, based heavily off the vignette for PBMC processing available on the amethyst github repo.\n",
        "\n",
        "We will be processing from breast cancer cell line cells.\n",
        "\n",
        "Our goals in this workshop are to:\n",
        "\n",
        "- Get familiar with quality metrics of DNA methylation\n",
        "- Cluster single cell methylomes to separate out the cell lines.\n",
        "- Perform copy number calling to confirm our clusters."
      ],
      "metadata": {
        "id": "u2Aw888UA5Tk"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Library Installation\n",
        "\n",
        "I prebuilt some R libraries that we can use to skip the install process. If you want to run the install yourself, the code is commented out below. You can fully ignore the commented out code below. All it is showing is that we need to install quite a few dependencies to get to the installation of two packages that will be the focus on this workshop:\n",
        "\n",
        "- **amethyst**: single cell methylation processing\n",
        "- **copykit**: single cell whole genome copy number calling"
      ],
      "metadata": {
        "id": "5rkLePAHBUL4"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#first lets install packages for the tutorial\n",
        "# this is prerun and stored on google drive to save time\n",
        "\n",
        "#install.packages(\"BiocManager\")\n",
        "\n",
        "#library(\"BiocManager\")\n",
        "#BiocManager::install(c(\"caret\", \"devtools\", \"data.table\", \"dplyr\", \"furrr\", \"future\", \"future.apply\",\n",
        "#  \"ggplot2\", \"grDevices\", \"gridExtra\", \"igraph\", \"irlba\", \"janitor\", \"Matrix\", \"methods\", \"pheatmap\",\n",
        "#  \"plotly\", \"plyr\", \"purrr\", \"randomForest\", \"rhdf5\", \"rtracklayer\", \"Rtsne\", \"scales\", \"stats\", \"stringr\",\n",
        "#  \"tibble\", \"tidyr\", \"umap\", \"utils\"))\n",
        "\n",
        "#library(\"devtools\")\n",
        "\n",
        "#devtools::install_github(\"JinmiaoChenLab/Rphenograph\")\n",
        "#devtools::install_github(\"KrishnaswamyLab/MAGIC/Rmagic\")\n",
        "#devtools::install_github(\"TomKellyGenetics/leiden\")\n",
        "#installing older version of amethyst for compatibility with cell line data files.\n",
        "#system('wget https://github.com/lrylaarsdam/amethyst/releases/download/v0.0.0.9000/amethyst_0.0.0.9000.tar.gz')\n",
        "#system('tar -xf amethyst_0.0.0.9000.tar.gz')\n",
        "#devtools::install(\"/content/amethyst\")\n",
        "#devtools::install_github(\"navinlabcode/copykit\")\n",
        "#system('cd /usr/local/lib/R/site-library/; tar -czf /content/R_library.tar.gz *; cd /content')\n",
        "\n"
      ],
      "metadata": {
        "id": "d8jGVU8PBf2S"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "To save ~30min of installation time, I prebuild the library and uploaded to a google drive folder. The following code downloads the libraries, and uncompresses them, so we can use them in our R session.\n",
        "\n",
        "We'll also download the cell line methylation data while here."
      ],
      "metadata": {
        "id": "zCim33NdCT6w"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "id": "-QpeUSLQ_od-"
      },
      "outputs": [],
      "source": [
        "#download prebuilt R libraries\n",
        "system(\"gdown '1r2P-GZa2QukEl-i619yK7dqeoEhGjBNN'\")\n",
        "system(\"mkdir R_library\")\n",
        "system(\"tar -xvf R_library.tar.gz -C R_library\") #unpack the library"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#download data\n",
        "system(\"gdown '195TI6-r2ruEGaKK4rIfWa24aT2xyEjL5'\")\n",
        "system(\"tar -xf workshop_data.tar.gz\") #unpack the library"
      ],
      "metadata": {
        "id": "ni42_iGg_EG6"
      },
      "execution_count": 6,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "If you check your files for this colab session, you should see them populate with two folders:\n",
        "\n",
        "- **R_library.tar.gz** : this is our downloaded compressed libraries\n",
        "- **/R_library/** : this is our decompressed folder\n",
        "- **h5_files** : this contains our methylation data in compressed h5 format\n",
        "- **metadata.tsv** : this is a tab-separated format of our metadata\n",
        "- **workshop.copykit.obj.rds** : this is a bonus copykit object that we will use for cnv calling\n",
        "\n",
        "Now all we have to do is tell R where to look for these libraries and we should be good to go!\n",
        "\n",
        "To do this, we are going to append the folder location to an R environment variable called \".libPaths\".\n",
        "\n",
        "Following this we will load our packages."
      ],
      "metadata": {
        "id": "M_BmbTmuHuL9"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        ".libPaths(c(\"/content/R_library\", .libPaths()))\n",
        "library(amethyst)\n",
        "library(copykit)\n",
        "library(ggplot2)\n",
        "\n",
        ".libPaths()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "VzVqptnMAEY8",
        "outputId": "841fcfaf-ce39-4a64-8e14-b8ce175a55bb"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“replacing previous import ‘GenomicRanges::reduce’ by ‘purrr::reduce’ when loading ‘amethyst’”\n",
            "Attaching amethyst version 0.0.0.9000.\n",
            "\n",
            "Loading required package: SingleCellExperiment\n",
            "\n",
            "Loading required package: SummarizedExperiment\n",
            "\n",
            "Loading required package: MatrixGenerics\n",
            "\n",
            "Loading required package: matrixStats\n",
            "\n",
            "\n",
            "Attaching package: ‘MatrixGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
            "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
            "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
            "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
            "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
            "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
            "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
            "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
            "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
            "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
            "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
            "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
            "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
            "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
            "    rowWeightedSds, rowWeightedVars\n",
            "\n",
            "\n",
            "Loading required package: GenomicRanges\n",
            "\n",
            "Loading required package: stats4\n",
            "\n",
            "Loading required package: BiocGenerics\n",
            "\n",
            "Loading required package: generics\n",
            "\n",
            "\n",
            "Attaching package: ‘generics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,\n",
            "    setequal, union\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘BiocGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    IQR, mad, sd, var, xtabs\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
            "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
            "    get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget,\n",
            "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
            "    rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique,\n",
            "    unsplit, which.max, which.min\n",
            "\n",
            "\n",
            "Loading required package: S4Vectors\n",
            "\n",
            "\n",
            "Attaching package: ‘S4Vectors’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:utils’:\n",
            "\n",
            "    findMatches\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    expand.grid, I, unname\n",
            "\n",
            "\n",
            "Loading required package: IRanges\n",
            "\n",
            "Loading required package: Seqinfo\n",
            "\n",
            "Loading required package: Biobase\n",
            "\n",
            "Welcome to Bioconductor\n",
            "\n",
            "    Vignettes contain introductory material; view with\n",
            "    'browseVignettes()'. To cite Bioconductor, see\n",
            "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘Biobase’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:MatrixGenerics’:\n",
            "\n",
            "    rowMedians\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    anyMissing, rowMedians\n",
            "\n",
            "\n",
            "Loading required package: Rsubread\n",
            "\n",
            "Loading required package: DNAcopy\n",
            "\n",
            "\n",
            "Attaching package: ‘copykit’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:amethyst’:\n",
            "\n",
            "    runUmap\n",
            "\n",
            "\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<style>\n",
              ".list-inline {list-style: none; margin:0; padding: 0}\n",
              ".list-inline>li {display: inline-block}\n",
              ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
              "</style>\n",
              "<ol class=list-inline><li>'/content/R_library'</li><li>'/usr/local/lib/R/site-library'</li><li>'/usr/lib/R/site-library'</li><li>'/usr/lib/R/library'</li></ol>\n"
            ],
            "text/markdown": "1. '/content/R_library'\n2. '/usr/local/lib/R/site-library'\n3. '/usr/lib/R/site-library'\n4. '/usr/lib/R/library'\n\n\n",
            "text/latex": "\\begin{enumerate*}\n\\item '/content/R\\_library'\n\\item '/usr/local/lib/R/site-library'\n\\item '/usr/lib/R/site-library'\n\\item '/usr/lib/R/library'\n\\end{enumerate*}\n",
            "text/plain": [
              "[1] \"/content/R_library\"            \"/usr/local/lib/R/site-library\"\n",
              "[3] \"/usr/lib/R/site-library\"       \"/usr/lib/R/library\"           "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Let's Assemble our Amethyst Object\n",
        "\n",
        "![amethyst_layout.png](https://github.com/lrylaarsdam/amethyst/blob/main/images/objectstructure.png?raw=true)"
      ],
      "metadata": {
        "id": "d8rGjJoWL6ZM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#read in our metadata first\n",
        "meta<-read.csv(file=\"/content/metadata.tsv\",header=T,sep=\"\\t\")\n",
        "head(meta)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 515
        },
        "id": "2yFbgcEvAnTB",
        "outputId": "d683268b-0055-4e22-eec7-008004ad5583"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A data.frame: 6 × 17</caption>\n",
              "<thead>\n",
              "\t<tr><th></th><th scope=col>cell_id</th><th scope=col>tgmt_well</th><th scope=col>i5_well</th><th scope=col>i7_well</th><th scope=col>plate</th><th scope=col>batch</th><th scope=col>plate_info</th><th scope=col>unique_reads</th><th scope=col>total_reads</th><th scope=col>mito_reads</th><th scope=col>mcg_pct</th><th scope=col>cov</th><th scope=col>ch_cov</th><th scope=col>tss_enrich</th><th scope=col>cg_cov</th><th scope=col>mch_pct</th><th scope=col>sample</th></tr>\n",
              "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><th scope=row>CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1</th><td>CGAATCTCCT+CTTCGGCGCA+AACTGTAG</td><td>1A06</td><td>E01</td><td>A05</td><td>plate1-1</td><td>prelim1-2</td><td>prelim1-2_scalebio_plate1-1</td><td>691748</td><td>1148393</td><td>0.001</td><td>68.92</td><td>12034132</td><td>11566095</td><td>0.7604</td><td>468037</td><td>0.68</td><td>MCF10A</td></tr>\n",
              "\t<tr><th scope=row>GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1</th><td>GTCTCCAGAG+TCGTAGGCTT+AACTGTAG</td><td>1A06</td><td>H01</td><td>A06</td><td>plate1-1</td><td>prelim1-2</td><td>prelim1-2_scalebio_plate1-1</td><td>640821</td><td>1037107</td><td>0.001</td><td>69.59</td><td>11440496</td><td>11008909</td><td>0.7406</td><td>431587</td><td>0.67</td><td>MCF10A</td></tr>\n",
              "\t<tr><th scope=row>GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1</th><td>GTCTCCAGAG+ACCGGAATTA+ATAATGTG</td><td>1B03</td><td>D01</td><td>A06</td><td>plate1-1</td><td>prelim1-2</td><td>prelim1-2_scalebio_plate1-1</td><td>555624</td><td> 905954</td><td>0.001</td><td>68.82</td><td> 9897578</td><td> 9523222</td><td>0.7230</td><td>374356</td><td>0.71</td><td>MCF10A</td></tr>\n",
              "\t<tr><th scope=row>AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1</th><td>AATTGAGAGA+TCGTAGGCTT+AACTGTAG</td><td>1A06</td><td>H01</td><td>A11</td><td>plate1-1</td><td>prelim1-2</td><td>prelim1-2_scalebio_plate1-1</td><td>536382</td><td> 873647</td><td>0.001</td><td>69.03</td><td> 9525741</td><td> 9162261</td><td>0.6952</td><td>363480</td><td>0.66</td><td>MCF10A</td></tr>\n",
              "\t<tr><th scope=row>TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1</th><td>TCGAAGCGCG+GGCAATGAGA+ATAATGTG</td><td>1B03</td><td>B01</td><td>A07</td><td>plate1-1</td><td>prelim1-2</td><td>prelim1-2_scalebio_plate1-1</td><td>497105</td><td> 776195</td><td>0.000</td><td>69.49</td><td> 9038954</td><td> 8701616</td><td>0.7497</td><td>337338</td><td>0.69</td><td>MCF10A</td></tr>\n",
              "\t<tr><th scope=row>CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1</th><td>CGAATCTCCT+GGCAATGAGA+AACTGTAG</td><td>1A06</td><td>B01</td><td>A05</td><td>plate1-1</td><td>prelim1-2</td><td>prelim1-2_scalebio_plate1-1</td><td>435953</td><td> 736252</td><td>0.000</td><td>70.15</td><td> 7644320</td><td> 7353858</td><td>0.7426</td><td>290462</td><td>0.70</td><td>MCF10A</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA data.frame: 6 × 17\n\n| <!--/--> | cell_id &lt;chr&gt; | tgmt_well &lt;chr&gt; | i5_well &lt;chr&gt; | i7_well &lt;chr&gt; | plate &lt;chr&gt; | batch &lt;chr&gt; | plate_info &lt;chr&gt; | unique_reads &lt;int&gt; | total_reads &lt;int&gt; | mito_reads &lt;dbl&gt; | mcg_pct &lt;dbl&gt; | cov &lt;int&gt; | ch_cov &lt;int&gt; | tss_enrich &lt;dbl&gt; | cg_cov &lt;int&gt; | mch_pct &lt;dbl&gt; | sample &lt;chr&gt; |\n|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n| CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 | CGAATCTCCT+CTTCGGCGCA+AACTGTAG | 1A06 | E01 | A05 | plate1-1 | prelim1-2 | prelim1-2_scalebio_plate1-1 | 691748 | 1148393 | 0.001 | 68.92 | 12034132 | 11566095 | 0.7604 | 468037 | 0.68 | MCF10A |\n| GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 | GTCTCCAGAG+TCGTAGGCTT+AACTGTAG | 1A06 | H01 | A06 | plate1-1 | prelim1-2 | prelim1-2_scalebio_plate1-1 | 640821 | 1037107 | 0.001 | 69.59 | 11440496 | 11008909 | 0.7406 | 431587 | 0.67 | MCF10A |\n| GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1 | GTCTCCAGAG+ACCGGAATTA+ATAATGTG | 1B03 | D01 | A06 | plate1-1 | prelim1-2 | prelim1-2_scalebio_plate1-1 | 555624 |  905954 | 0.001 | 68.82 |  9897578 |  9523222 | 0.7230 | 374356 | 0.71 | MCF10A |\n| AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 | AATTGAGAGA+TCGTAGGCTT+AACTGTAG | 1A06 | H01 | A11 | plate1-1 | prelim1-2 | prelim1-2_scalebio_plate1-1 | 536382 |  873647 | 0.001 | 69.03 |  9525741 |  9162261 | 0.6952 | 363480 | 0.66 | MCF10A |\n| TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1 | TCGAAGCGCG+GGCAATGAGA+ATAATGTG | 1B03 | B01 | A07 | plate1-1 | prelim1-2 | prelim1-2_scalebio_plate1-1 | 497105 |  776195 | 0.000 | 69.49 |  9038954 |  8701616 | 0.7497 | 337338 | 0.69 | MCF10A |\n| CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1 | CGAATCTCCT+GGCAATGAGA+AACTGTAG | 1A06 | B01 | A05 | plate1-1 | prelim1-2 | prelim1-2_scalebio_plate1-1 | 435953 |  736252 | 0.000 | 70.15 |  7644320 |  7353858 | 0.7426 | 290462 | 0.70 | MCF10A |\n\n",
            "text/latex": "A data.frame: 6 × 17\n\\begin{tabular}{r|lllllllllllllllll}\n  & cell\\_id & tgmt\\_well & i5\\_well & i7\\_well & plate & batch & plate\\_info & unique\\_reads & total\\_reads & mito\\_reads & mcg\\_pct & cov & ch\\_cov & tss\\_enrich & cg\\_cov & mch\\_pct & sample\\\\\n  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <int> & <dbl> & <dbl> & <int> & <int> & <dbl> & <int> & <dbl> & <chr>\\\\\n\\hline\n\tCGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2\\_scalebio\\_plate1-1 & CGAATCTCCT+CTTCGGCGCA+AACTGTAG & 1A06 & E01 & A05 & plate1-1 & prelim1-2 & prelim1-2\\_scalebio\\_plate1-1 & 691748 & 1148393 & 0.001 & 68.92 & 12034132 & 11566095 & 0.7604 & 468037 & 0.68 & MCF10A\\\\\n\tGTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2\\_scalebio\\_plate1-1 & GTCTCCAGAG+TCGTAGGCTT+AACTGTAG & 1A06 & H01 & A06 & plate1-1 & prelim1-2 & prelim1-2\\_scalebio\\_plate1-1 & 640821 & 1037107 & 0.001 & 69.59 & 11440496 & 11008909 & 0.7406 & 431587 & 0.67 & MCF10A\\\\\n\tGTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2\\_scalebio\\_plate1-1 & GTCTCCAGAG+ACCGGAATTA+ATAATGTG & 1B03 & D01 & A06 & plate1-1 & prelim1-2 & prelim1-2\\_scalebio\\_plate1-1 & 555624 &  905954 & 0.001 & 68.82 &  9897578 &  9523222 & 0.7230 & 374356 & 0.71 & MCF10A\\\\\n\tAATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2\\_scalebio\\_plate1-1 & AATTGAGAGA+TCGTAGGCTT+AACTGTAG & 1A06 & H01 & A11 & plate1-1 & prelim1-2 & prelim1-2\\_scalebio\\_plate1-1 & 536382 &  873647 & 0.001 & 69.03 &  9525741 &  9162261 & 0.6952 & 363480 & 0.66 & MCF10A\\\\\n\tTCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2\\_scalebio\\_plate1-1 & TCGAAGCGCG+GGCAATGAGA+ATAATGTG & 1B03 & B01 & A07 & plate1-1 & prelim1-2 & prelim1-2\\_scalebio\\_plate1-1 & 497105 &  776195 & 0.000 & 69.49 &  9038954 &  8701616 & 0.7497 & 337338 & 0.69 & MCF10A\\\\\n\tCGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2\\_scalebio\\_plate1-1 & CGAATCTCCT+GGCAATGAGA+AACTGTAG & 1A06 & B01 & A05 & plate1-1 & prelim1-2 & prelim1-2\\_scalebio\\_plate1-1 & 435953 &  736252 & 0.000 & 70.15 &  7644320 &  7353858 & 0.7426 & 290462 & 0.70 & MCF10A\\\\\n\\end{tabular}\n",
            "text/plain": [
              "                                                           cell_id                       \n",
              "CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 CGAATCTCCT+CTTCGGCGCA+AACTGTAG\n",
              "GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 GTCTCCAGAG+TCGTAGGCTT+AACTGTAG\n",
              "GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1 GTCTCCAGAG+ACCGGAATTA+ATAATGTG\n",
              "AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 AATTGAGAGA+TCGTAGGCTT+AACTGTAG\n",
              "TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1 TCGAAGCGCG+GGCAATGAGA+ATAATGTG\n",
              "CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1 CGAATCTCCT+GGCAATGAGA+AACTGTAG\n",
              "                                                           tgmt_well i5_well\n",
              "CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 1A06      E01    \n",
              "GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 1A06      H01    \n",
              "GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1 1B03      D01    \n",
              "AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 1A06      H01    \n",
              "TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1 1B03      B01    \n",
              "CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1 1A06      B01    \n",
              "                                                           i7_well plate   \n",
              "CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 A05     plate1-1\n",
              "GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 A06     plate1-1\n",
              "GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1 A06     plate1-1\n",
              "AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 A11     plate1-1\n",
              "TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1 A07     plate1-1\n",
              "CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1 A05     plate1-1\n",
              "                                                           batch    \n",
              "CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 prelim1-2\n",
              "GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 prelim1-2\n",
              "GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1 prelim1-2\n",
              "AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 prelim1-2\n",
              "TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1 prelim1-2\n",
              "CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1 prelim1-2\n",
              "                                                           plate_info                 \n",
              "CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 prelim1-2_scalebio_plate1-1\n",
              "GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 prelim1-2_scalebio_plate1-1\n",
              "GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1 prelim1-2_scalebio_plate1-1\n",
              "AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 prelim1-2_scalebio_plate1-1\n",
              "TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1 prelim1-2_scalebio_plate1-1\n",
              "CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1 prelim1-2_scalebio_plate1-1\n",
              "                                                           unique_reads\n",
              "CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 691748      \n",
              "GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 640821      \n",
              "GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1 555624      \n",
              "AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 536382      \n",
              "TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1 497105      \n",
              "CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1 435953      \n",
              "                                                           total_reads\n",
              "CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 1148393    \n",
              "GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 1037107    \n",
              "GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1  905954    \n",
              "AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1  873647    \n",
              "TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1  776195    \n",
              "CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1  736252    \n",
              "                                                           mito_reads mcg_pct\n",
              "CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 0.001      68.92  \n",
              "GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 0.001      69.59  \n",
              "GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1 0.001      68.82  \n",
              "AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 0.001      69.03  \n",
              "TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1 0.000      69.49  \n",
              "CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1 0.000      70.15  \n",
              "                                                           cov      ch_cov  \n",
              "CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 12034132 11566095\n",
              "GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 11440496 11008909\n",
              "GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1  9897578  9523222\n",
              "AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1  9525741  9162261\n",
              "TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1  9038954  8701616\n",
              "CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1  7644320  7353858\n",
              "                                                           tss_enrich cg_cov\n",
              "CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 0.7604     468037\n",
              "GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 0.7406     431587\n",
              "GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1 0.7230     374356\n",
              "AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 0.6952     363480\n",
              "TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1 0.7497     337338\n",
              "CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1 0.7426     290462\n",
              "                                                           mch_pct sample\n",
              "CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 0.68    MCF10A\n",
              "GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 0.67    MCF10A\n",
              "GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1 0.71    MCF10A\n",
              "AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 0.66    MCF10A\n",
              "TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1 0.69    MCF10A\n",
              "CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1 0.70    MCF10A"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#read in our h5 filesobj <- addCellInfo(obj, file = \"metadata.tsv\")\n",
        "h5_paths<-read.csv(file=\"/content/h5_files/h5_paths.tsv\",header=T,sep=\"\\t\")\n",
        "h5_paths$barcodes<-row.names(h5_paths)\n",
        "h5_paths$paths<-h5_paths$path\n",
        "h5_paths<-h5_paths[,c(\"barcodes\",\"paths\")]\n",
        "head(h5_paths)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 410
        },
        "id": "rUKreygCMkub",
        "outputId": "954445b2-4eb5-4bba-a1e7-acce220b2917"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A data.frame: 6 × 2</caption>\n",
              "<thead>\n",
              "\t<tr><th></th><th scope=col>barcodes</th><th scope=col>paths</th></tr>\n",
              "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><th scope=row>CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1</th><td>CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1</td><td>/content/h5_files/MCF10A.1A06_cov.h5</td></tr>\n",
              "\t<tr><th scope=row>GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1</th><td>GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1</td><td>/content/h5_files/MCF10A.1A06_cov.h5</td></tr>\n",
              "\t<tr><th scope=row>GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1</th><td>GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1</td><td>/content/h5_files/MCF10A.1B03_cov.h5</td></tr>\n",
              "\t<tr><th scope=row>AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1</th><td>AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1</td><td>/content/h5_files/MCF10A.1A06_cov.h5</td></tr>\n",
              "\t<tr><th scope=row>TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1</th><td>TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1</td><td>/content/h5_files/MCF10A.1B03_cov.h5</td></tr>\n",
              "\t<tr><th scope=row>CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1</th><td>CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1</td><td>/content/h5_files/MCF10A.1A06_cov.h5</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA data.frame: 6 × 2\n\n| <!--/--> | barcodes &lt;chr&gt; | paths &lt;chr&gt; |\n|---|---|---|\n| CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 | CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 | /content/h5_files/MCF10A.1A06_cov.h5 |\n| GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 | GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 | /content/h5_files/MCF10A.1A06_cov.h5 |\n| GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1 | GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1 | /content/h5_files/MCF10A.1B03_cov.h5 |\n| AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 | AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 | /content/h5_files/MCF10A.1A06_cov.h5 |\n| TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1 | TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1 | /content/h5_files/MCF10A.1B03_cov.h5 |\n| CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1 | CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1 | /content/h5_files/MCF10A.1A06_cov.h5 |\n\n",
            "text/latex": "A data.frame: 6 × 2\n\\begin{tabular}{r|ll}\n  & barcodes & paths\\\\\n  & <chr> & <chr>\\\\\n\\hline\n\tCGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2\\_scalebio\\_plate1-1 & CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2\\_scalebio\\_plate1-1 & /content/h5\\_files/MCF10A.1A06\\_cov.h5\\\\\n\tGTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2\\_scalebio\\_plate1-1 & GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2\\_scalebio\\_plate1-1 & /content/h5\\_files/MCF10A.1A06\\_cov.h5\\\\\n\tGTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2\\_scalebio\\_plate1-1 & GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2\\_scalebio\\_plate1-1 & /content/h5\\_files/MCF10A.1B03\\_cov.h5\\\\\n\tAATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2\\_scalebio\\_plate1-1 & AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2\\_scalebio\\_plate1-1 & /content/h5\\_files/MCF10A.1A06\\_cov.h5\\\\\n\tTCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2\\_scalebio\\_plate1-1 & TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2\\_scalebio\\_plate1-1 & /content/h5\\_files/MCF10A.1B03\\_cov.h5\\\\\n\tCGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2\\_scalebio\\_plate1-1 & CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2\\_scalebio\\_plate1-1 & /content/h5\\_files/MCF10A.1A06\\_cov.h5\\\\\n\\end{tabular}\n",
            "text/plain": [
              "                                                           barcodes                                                  \n",
              "CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1\n",
              "GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1\n",
              "GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1 GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1\n",
              "AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1\n",
              "TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1 TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1\n",
              "CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1 CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1\n",
              "                                                           paths                               \n",
              "CGAATCTCCT+CTTCGGCGCA+AACTGTAG+prelim1-2_scalebio_plate1-1 /content/h5_files/MCF10A.1A06_cov.h5\n",
              "GTCTCCAGAG+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 /content/h5_files/MCF10A.1A06_cov.h5\n",
              "GTCTCCAGAG+ACCGGAATTA+ATAATGTG+prelim1-2_scalebio_plate1-1 /content/h5_files/MCF10A.1B03_cov.h5\n",
              "AATTGAGAGA+TCGTAGGCTT+AACTGTAG+prelim1-2_scalebio_plate1-1 /content/h5_files/MCF10A.1A06_cov.h5\n",
              "TCGAAGCGCG+GGCAATGAGA+ATAATGTG+prelim1-2_scalebio_plate1-1 /content/h5_files/MCF10A.1B03_cov.h5\n",
              "CGAATCTCCT+GGCAATGAGA+AACTGTAG+prelim1-2_scalebio_plate1-1 /content/h5_files/MCF10A.1A06_cov.h5"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "You'll notice for both the h5 paths and the metadata, the row names of the files are the same. So through this we will like our cell info to the methylation profiles! Now let's make a new amethyst object."
      ],
      "metadata": {
        "id": "sEzIsz_EZdi1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "obj <- createObject()\n",
        "str(obj)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ro7fnFXfZmaB",
        "outputId": "8152723b-e20f-48ea-8dda-0e8acd5a24e0"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Formal class 'amethyst' [package \"amethyst\"] with 6 slots\n",
            "  ..@ h5paths       : NULL\n",
            "  ..@ genomeMatrices: NULL\n",
            "  ..@ reductions    : NULL\n",
            "  ..@ index         : NULL\n",
            "  ..@ metadata      : NULL\n",
            "  ..@ ref           : NULL\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "The \"@\" symbol shows a slot in the object. These are predefined in R, so they only accept certain data types and formats.\n",
        "\n",
        "We will start by adding our h5 paths into the @h5paths slot, and our metadata into the @metadata slot. We'll then just do another quick check to make sure it got added correctly."
      ],
      "metadata": {
        "id": "BIyA-t2ZZufn"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "obj@metadata<-meta\n",
        "obj@h5paths<-h5_paths[row.names(meta),]\n"
      ],
      "metadata": {
        "id": "gaA90NcOZ8ud"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "R structured objects are hierarchical. So we can call a slot, and then call further objected within that slot.\n",
        "\n",
        "Let's make a quick counts table of our \"sample\" metadata column."
      ],
      "metadata": {
        "id": "PMe_rFIjaScL"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "table(obj@metadata$sample)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 71
        },
        "id": "eQdi4F_7aSzP",
        "outputId": "1a28896f-ea91-4f13-aea8-b468023d25c4"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "\n",
              "    MCF10A       MCF7 MDA-MB-231 \n",
              "        55         52         48 "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Looks like we're working with about 50 cells a piece for 3 different cell lines. Let's take a peek at the coverage per cell as a histogram."
      ],
      "metadata": {
        "id": "hGi1iK0Pa4IM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "ggplot(obj@metadata, aes(x = cov)) + geom_histogram(bins = 10) + theme_classic()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 437
        },
        "id": "2tsswidpavli",
        "outputId": "dbfe1ae8-8db2-4e8c-c22e-cc9e5a109424"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "plot without title"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAClFBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0PDw8QEBARERESEhITExMUFBQVFRUWFhYX\nFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8hISEiIiIjIyMkJCQmJiYnJycoKCgpKSkrKyss\nLCwtLS0uLi4vLy8xMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/\nPz9AQEBBQUFCQkJDQ0NERERGRkZHR0dISEhJSUlKSkpLS0tOTk5PT09QUFBRUVFSUlJTU1NV\nVVVWVlZXV1dYWFhZWVlaWlpbW1tdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZoaGhp\naWlqampra2ttbW1ubm5vb29xcXFycnJzc3N0dHR1dXV2dnZ3d3d5eXl+fn5/f3+AgICBgYGC\ngoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uNjY2Ojo6Pj4+RkZGSkpKTk5OVlZWXl5eZ\nmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKiqqqqsrKyt\nra2urq6vr6+wsLCxsbGzs7O0tLS1tbW2tra3t7e4uLi6urq7u7u8vLy9vb2+vr6/v7/BwcHD\nw8PFxcXGxsbIyMjJycnKysrLy8vMzMzOzs7Pz8/Q0NDS0tLT09PV1dXW1tbX19fY2NjZ2dna\n2trb29vd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt\n7e3u7u7v7+/x8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///+3\nx5qhAAAACXBIWXMAABJ0AAASdAHeZh94AAAYjUlEQVR4nO3d/9/kZV3o8blZMnaRBBZwFzJW\nK76YCFq5Ua1HIyjAiihSizwGaacyKaKv54idNdS+2BcLKCCBkjJZYysULHT3LJt8C+RY1rbz\nzzRzz67Mxwc6c332/Xlf98U8nz/M3DMPP4+5rvf1eT323r2XdTQGjtmo9gLghUBIEEBIEEBI\nEEBIEEBIEEBIEKBnSP/+P94Zuw5oWs+QvjD67th1QNOEBAGEBAGEBAGEBAGEBAGEBAGEBAGE\nBAGEBAGEBAGEBAGEBAGEBAGEBAGEBAGEBAGEBAGEBAGEBAGEBAGEBAGEBAGEBAGEBAGEBAGE\nBAGEBAGEBAGEBAGEBAGEBAGEBAGEBAGEBAGEBAEaCOmqEGnLZSUJCQIICQIICQIICQIICQII\nCQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQII\nCQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQII\nCQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQII\nCQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQIICQII\nCQIICQIICQIICQIICQIICQIICQIICQIICQIsEdJvnf2i8/98PH7q6u2nXPrIkfeEBPMWh/TB\nl97xmf/98qfHl+3c+/BV5x6avSkkmLc4pJf/zvrTvrUHJr8qHX/P7E0hwbyFIf2/0e+8cstr\nPja+dfPhyavzb5y9KySYtzCkvx195yefvP4lj+0+c/pq13WThx84+eSXCAnmLBHS5Lu5/zj5\ng7vPmr7adf3k4S07dpwtJJizMKR9oz2Tx3Nvuv2E6bd25900e9e3djBvYUiHtt086eakPzqw\nNgnq8U33zd4VEsxb/Kd2v7L17n1v3fbs+MoL9j50yUWHZ28KCeYtDunQz5/xoosfHI+fvmbb\n1ssfPfKmkGCevyIEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQE\nAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQE\nAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQE\nAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQE\nAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQE\nAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQE\nAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQE\nAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQE\nAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQE\nAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQE\nAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQEAYQE\nAZYK6YOj28bjp67efsqljxx5R0gwb5mQ/uWMzZOQLtu59+Grzj00e0tIMG+ZkK58xxm3jfet\nPTD5Ven4e2ZvCQnmLRHSn5797CSkWzcfnnx9/o2z94QE8xaH9NS2u8eTkHafOX2x67rJw83X\nXvsWIcGcxSH92I+N10M6a/pi1/WTh+8dTbQWUoy0TdOYhSHd/dIn10O6/YTpt3bn3TR5+PQn\nPvE3QoI5C0P6oRNOPfXUtZOuOLC2Zzx+fNN9s3fb+z1SjLRN05iFIT25f+K0Wx4fX3nB3ocu\nuejw7F0hwbzl/mbD5Fu78dPXbNt6+aNH3hASzFuZvyIUI23TNEZIRdI2TWOEVCRt0zRGSEXS\nNk1jhFQkbdM0RkhF0jZNY4RUJG3TNEZIRdI2TWOEVCRt0zRGSEXSNk1jhFQkbdM0RkhF0jZN\nY4RUJG3TNEZIRdI2TWOEVCRt0zRGSEXSNk1jhFQkbdM0RkhF0jZNY4RUJG3TNEZIRdI2TWOE\nVCRt0zRGSEXSNk1jhFQkbdM0RkhF0jZNY4RUJG3TNEZIRdI2TWOEVCRt0zRGSEXSNk1jhFQk\nbdM0RkhF0jZNYzohvfrB2fOfnLPoMiHBvE5Ioz3rT/95w4sWXSYkmDcX0ug5Fyy6TEgwby6k\nve8Zfd9bp378XfsXXSYkmNf51u6NDy97mZBgnj+1K5K2aRrTCelz12w/bvabpEWXCQnmdZJ5\n0/Gvv2b9d0lvXXSZkGBeJ6RTb1/2MiHBvE5IWx5b9jIhwbxOSN/x0WUvExLM64T0idd8bMnL\nhATzOiG97utHW75h3aLLhATzut/avf6oRZcJCeb5gWyRtE3TGCEVSds0jen+HOmokxZdJiSY\n1wnp+9a9ZvN5b190mZBg3vN9a3fw4jsWXSYkmPe8v0fa8+pFlwkJ5j1vSAc3L7pMSDDv+UI6\n/MtnLbpMSDCvE9K3rDtv6+hnFl0mJJj3PCG9atd7vrjoMiHBPD+QLZK2aRrzZSE9ccf7bvnI\nM4svExLM64T0X+/8muk/2HDiry+8TEgwrxPSr48uf/9dd+x+4+h3F10mJJjXCemcd8yer/Uv\nrX4FaZumMZ2QvvYvZ893+oHsV5C2aRrTCenEP5893/7iRZcJCeZ1Qtr53es/QPq3N3zXosuE\nBPM6Id259rKf/KVf/Intx92z6DIhwbzuz5Fu++bpH3+ff+fCy4QE8778bzYcuH/PvyxxmZBg\nXjekgzdPHh674XMLLxMSzOuE9KmXTv8/Lz8zeuk/LbpMSDCvE9L3v+L+6dODr7hi0WVCgnmd\nkE77wOx5t39F6CtI2zSN6YS0+UOz5z/YsugyIcG8Tkjf/sZD06dnLnrdosuEBPM6IX1kbcfb\n3/2uN5923EcWXSYkmNf94++7Xz39gewr71x4mZBg3pf/QPaJv//HJf4DWSFBh3+zoUjapmmM\nkIqkbZrGCKlI2qZpjJCKpG2axgipSNqmaYyQiqRtmsYIqUjapmmMkIqkbZrGCKlI2qZpjJCK\npG2axgipSNqmaYyQiqRtmsYIqUjapmmMkIqkbZrGCKlI2qZpjJCKpG2axgipSNqmaYyQiqRt\nmsYIqUjapmmMkIqkbZrGCKlI2qZpjJCKpG2axgipSNqmaYyQiqRtmsYIqUjapmmMkIqkbZrG\nCKlI2qZpjJCKpG2axgipSNqmaYyQiqRtmsYIqUjapmmMkIqkbZrGCKlI2qZpjJCKpG2axgip\nSNqmaYyQiqRtmsYIqUjapmmMkIqkbZrGCKlI2qZpjJCKpG2axgipSNqmaYyQiqRtmsYIqUja\npmmMkIqkbZrGCKlI2qZpjJCKpG2axgipSNqmaYyQiqRtmsYIqUjapmmMkIqkbZrGCKlI2qZp\njJCKpG2axgipSNqmaYyQiqRtmsYIqUjapmmMkIqkbZrGCKlI2qZpjJCKpG2axgipSNqmaYyQ\niqRtmsYIqUjapmmMkIqkbZrGCKlI2qZpjJCKpG2axgipSNqmaYyQiqRtmsYIqUjapmmMkIqk\nbZrGCKlI2qZpjJCKpG2axgipSNqmaYyQiqRtmsYIqUjapmmMkIqkbZrGCKlI2qZpjJCKpG2a\nxgipSNqmaYyQiqRtmsYIqUjapmmMkIqkbZrGCKlI2qZpjJCKpG2axgipSNqmaYyQiqRtmsYI\nqUjapmmMkIqkbZrGCKmGtNmRRUg1pM2OLEKqIW12ZBFSDWmzI4uQakibHVmEVEPa7MgipBrS\nZkcWIdWQNjuyCKmGtNmRRUg1pM2OLEKqIW12ZBFSDWmzI4uQakibHVmEVEPa7MgipBrSZkcW\nIdWQNjuyCKmGtNmRRUg1pM2OLEKqIW12ZBFSDWmzI4uQakibHVmEVEPa7MgipBrSZkcWIdWQ\nNjuyCKmGtNmRRUg1pM2OLEKqIW12ZBFSDWmzI4uQakibHVmEVEPa7MgipBrSZkcWIdWQNjuy\nCKmGtNmRRUg1pM2OLEKqIW12ZBFSDWmzI8vikA5cdfpJF398PH7q6u2nXPrIkTeFdEzSZkeW\nxSFduPPvPv3DW58dX7Zz78NXnXto9qaQjkna7MiyMKQnr3hwPP7s6P59aw9MflU6/p7Zu0I6\nJmmzI8tyv0f62KaDt24+PPni/BsnDw/cc88dQjoWabMjy1IhPXnOz453nzn9atd1k4fvHU0I\n6RikzY4sy4T0yVe87fB491nTL3ddP3n4wM/93DuFdCzSZkeWJUK699SbJ4+3nzD91u68m2bv\n+T3SMUmbHVkWh/TXJ981fTqwtmc8fnzTfbM3hXRM0mZHloUhfWHHDfsnnh1fecHehy656PCR\nd4V0LNJmR5aFId07Wvfe8dPXbNt6+aNH3hXSMUmbHVn8FaEa0mZHFiHVkDY7sgiphrTZkUVI\nNaTNjixCqiFtdmQRUg1psyOLkGpImx1ZhFRD2uzIIqQa0mZHFiHVkDY7sgiphrTZkUVINaTN\njixCqiFtdmQRUg1psyOLkGpImx1ZhFRD2uzIIqQa0mZHFiHVkDY7sgiphrTZkUVINaTNjixC\nqiFtdmQRUg1psyOLkGpImx1ZhFRD2uzIIqQa0mZHFiHVkDY7sgiphrTZkUVINaTNjixCqiFt\ndmQRUg1psyOLkGpImx1ZhFRD2uzIIqQa0mZHFiHVkDY7sgiphrTZkUVINaTNjixCqiFtdmQR\nUg1psyOLkGpImx1ZhFRD2uzIIqQa0mZHFiHVkDY7sgiphrTZkUVINaTNjixCqiFtdmQRUg1p\nsyOLkGpImx1ZhFRD2uzIIqQa0mZHFiHVkDY7sgiphrTZkUVINaTNjixCqiFtdmQRUg1psyOL\nkGpImx1ZhFRD2uzIIqQa0mZHFiHVkDY7sgiphrTZkUVINaTNjixCqiFtdmQRUg1psyOLkGpI\nmx1ZhFRD2uzIIqQa0mZHFiHVkDY7sgiphrTZkUVINaTNjixCqiFtdmQRUg1psyOLkGpImx1Z\nhFRD2uzIIqQa0mZHFiHVkDY7sgiphrTZkUVINaTNjixCqiFtdmQRUg1psyOLkGpImx1ZhFRD\n2uzIIqQa0mZHFiHVkDY7sgiphrTZkUVINaTNjixCqiFtdmQRUg1psyOLkGpImx1ZhFRD2uzI\nIqQa0mZHFiHVkDY7sgiphrTZkUVINaTNjixCqiFtdmQRUg1psyOLkGpImx1ZhFRD2uzIIqQa\n0mZHFiHVkDY7sgiphrTZkUVINaTNjixCqiFtdmQRUg1psyOLkGpImx1ZhFRD2uzIIqQa0mZH\nFiHVkDY7sgiphrTZkUVINaTNjixCalfaCbCYkNqVdgIsJqR2pZ0AiwmpXWknwGJCalfaCbCY\nkNqVdgIsJqR2pZ0AiwmpXWknwGJCalfaCbCYkNqVdgIsNmRItW+0F7p+R8cghNSufkfHIITU\nrn5HxyCE1K5+R8cghNSufkfHIITUrn5HxyCE1K5+R8cghNSufkfHIITUrn5HxyCE1K5+R8cg\nhNSufkfHIITUrn5HxyCE1K5+R8cghNSufkfHIITUrn5HxyCE1K5+R8cghNSufkfHIITUrn5H\nxyCE1K5+R8cghNSufkfHIITUrn5HxyCE1K5+R8cghNSufkfHIITUrn5HxyCE1K5+R8cghNSu\nfkfHIITUrn5HxyCE1K5+R8cghNSufkfHIITUrn5HxyCE1K5+R8cghNSufkfHIITUrn5HxyCE\n1K5+R8cghNSufkfHIITUrn5HxyCE1K5+R8cghNSufkfHIITUrn5HxyCE1K5+R8cghNSufkfH\nIITUrn5HxyCE1K5+R8cghNSufkfHIITUrn5HxyCE1K5+R8cghNSufkfHIITUrn5HxyCE1K5+\nR8cghNSufkfHIITUrn5HxyCEtPL63QF0CWnl9bsD6BLSyut3B9AlpJXX7w6gS0grr98dQJeQ\nVl6/O4AuIa28fncAXUJaef3uALqEtPL63QF0CWnl9bsD6BLSyut3B9AlpJXX7w6gS0grr98d\nQJeQVl6/O4AuIa28fncAXUJaef3uALqEtPL63QF0CWnl9bsD6BLSyut3B9AlpJXX7w6gS0gr\nr98dQJeQVl6/O2AgtYfRUbJwIa28fnfAQGoPo6Nk4UJaef3ugIHUHkZHycKFtPL63QEDqT2M\njpKFC2nl9bsDBlJ7GB0lC18+pKeu3n7KpY8ceSGkF46S22VwtYfRUbLw5UO6bOfeh68699Ds\nhZBeOEpul8HVHkZHycKXDmnf2gOTX5WOv2f2SkgvHCW3y+BqD6OjZOFLh3Tr5sOTx/NvnL0S\n0gtHye0yuNrD6ChZ+NIh7T5z+rjrusnDD5x88kuE9IJRcrsMrvYwOkoWvnxIZ00fd10/eXjL\njh1nLxMSrIylQ7r9hOm3dufdNHu11Ld2sDKWDunA2p7x+PFN981eCQnmLf/H31desPehSy46\nPHshJJi3fEhPX7Nt6+WPHnkhJJg35F8RgpUhJAggJAggJAggJAggJAggJAggJAggJAggJAgg\nJAggJAggJAggJAggJAggJAggJAggJAggJAggJAggJAggJAggJAggJAggJAggJAggJAggJAgg\nJAjQM6RnRi//VVh5f3CMIT05AkYXH2NIX/zpd3+43IWj3+5x1WD+7+jbai+h4/LRL9ReQseW\nM2uvoOOnRm+uvYSObxr98Yc//NFjDKmfy0afy/y4RfaN3lR7CR0/P/qr2kvo+Lpzaq+g40Oj\nm2svoeN1o0Nzr4S0cQjpqxLSlwjpqxLSVyUkeMETEgQQEgQQEgRIDOmpq7efcukjeZ931Cun\nPzg78bnPX/Z5CJ967abpU+lSBlvSkfVslBEduOr0ky7++IaZz9HlLDOexJAu27n34avOPbT4\nfxjsrJv3799/4LnPX/Z5AH+87UfXb9zSpQy1pKPr2SgjunDn3336h7c+u1Hmc3Q5y4wnL6R9\naw9M2j3+nrQPPGrLnZ3PX/Z5iKX87mdv29RjKYMt6ch6NsqInrziwfH4s6P7N8h8ji5nqfHk\nhXTr5sOTx/NvTPvAI/599NZXveyKh770+cs+D7Oa9Ru3dCkDLml9PRtqRB/bdHADzWe6nKXG\nkxfS7jOnj7uuS/vAIx4740c+/reXnPGvRz9/2edhVrN+45YuZcAlra9nI43oyXN+diPNZ7qc\npcaTGNJZ6x97fdoHznvmxFuOfv6yz8MsZBZS4VIGXNLsW7upjTGiT77ibYc30HzWl7Nu0Xjy\nQrr9hOmSzrsp7QM7zvmFo5+/7PMw61i/cUuXMuCSngtpQ4zo3lOnfw1ow8xntpyZBePJC+nA\n2p7x+PFN96V94BH/8ONfHI8/f+LvHf38ZZ+HWc36jVu6lAGXtL6eDTOivz75runTRpnPkeUs\nNZ7EP/6+8oK9D11y0eG8D5x54tQf/adPXfH1//9Ln7/s8wAO7r9l0/79ny9eylBLOrKejTKi\nL+y4Yf/EsxtkPkeXs9R4EkN6+pptWy9/NO/zjnrg9V932mX//NznL/s8gG9Y/68q/0/xUoZa\n0tH1bJAR3Tv7r07fu0Hm86XlLDMef0UIAggJAggJAggJAggJAggJAggJAggJAggJAggJAggJ\nAgipLXdf/OIz3vTpyRd3fceLTzj3Nw+PX3fqf07ff+22/H8MgzlCasrda2/40Pt3bDs4vm3t\ne26/9x2j/zX+rdHdk/c/u/aO2ktbcUJqyoVnT379+fiL3jP+5pd9cfLy+7/micePv3byxW+M\nHqi9tBUnpJY8Mfqfsy8OjH5y+vT+0R3jS07/r/H4onNrLgshteUfRu+efXH/6JemT3eN3jf+\n/dFHx4+MfrXmshBSW/5x9K7ZF3tGN0yf7hzdMv78lrePf21tX81lIaS2PDN68/TpM48dHE1/\nZzR+3+gj4/EPbh9f+F1Vl4WQGnP+ac+Mx5+cfIN33vZ/m7z8ni1Pj8d/Nrpt9P7aC1t5QmrK\nHce99g/f942nHxzfedwb/uwv3jb6lcl7/3HKjhOerr2wlSekttz5rVtOv/zhyRd37zzxa1/1\ngfX3rt1g/xeeK0lIEEBIEEBIEEBIEEBIEEBIEEBIEEBIEEBIEEBIEEBIEEBIEOC/AST++jxR\n1UAOAAAAAElFTkSuQmCC"
          },
          "metadata": {
            "image/png": {
              "width": 420,
              "height": 420
            }
          }
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Clustering\n",
        "\n",
        "Now let's cluster our data based on the methylome profiles. First we will index our CG sites in our h5 files. This will allow us to access them for summarizing over windows more efficiently. We can now take a peek at some chr1 data."
      ],
      "metadata": {
        "id": "xgu30UJFbEVD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "obj@index[[\"chr_cg\"]] <- indexChr(obj, type = \"CG\", threads = 4)\n",
        "head(obj@index[[\"chr_cg\"]][[\"chr1\"]])"
      ],
      "metadata": {
        "id": "B2tuyTDs8tDc"
      },
      "execution_count": 15,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "obj@genomeMatrices[[\"cg_100k_score\"]] <- makeWindows(obj,\n",
        "                                                    stepsize = 100000,\n",
        "                                                    type = \"CG\",\n",
        "                                                    metric = \"score\",\n",
        "                                                    threads = 1,\n",
        "                                                    index = \"chr_cg\",\n",
        "                                                    nmin = 2)"
      ],
      "metadata": {
        "id": "JB7G0mJk_Q7a"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}